\documentclass[12pt]{article}

\usepackage[utf8]{inputenc} % declare encoding as utf8
\usepackage{graphicx} % to enable \includegraphics
\usepackage{flafter} % to make sure floats appear after their position in text
\usepackage{pdflscape} % allow some pages to be in landscape orientation
\usepackage{layout} % for debugging purposes, looking at page layout
\usepackage[margin=1in]{geometry} % leaves more space for text body
\usepackage[hidelinks]{hyperref} % hyperlinks from list of figures/tables
\usepackage[labelfont=bf]{caption} % format captions and make links go to figure instead of caption
\usepackage{csvsimple} % read and display csv files as tables
\usepackage{booktabs} % prettier tables
\usepackage{threeparttable} % easy addition of footnotes below tables
\usepackage{textcomp} % makes the \textdegree command available
\usepackage[numbers]{natbib} % bibliography commands
\usepackage{tocloft} % formatting the lists of tables and figures
\usepackage{xcolor} % enable text coloring
\usepackage{calc} % makes \widthof available to set the width of the last column
\usepackage{array} % using command \arraybackslash to restore the definition of \\ after \centering
\usepackage{longtable} % allow tables to span more than one page
\usepackage{threeparttablex} % threeparttable for longtable
\usepackage{capt-of} % put caption on next page

% Formatting the table of contents/subsection
\renewcommand{\contentsname}{Supplemental methods}

% Formatting the list of tables
\renewcommand{\listtablename}{Supplemental tables} % Write "Supplemental tables" instead of "List of tables"
\renewcommand\cfttabpresnum{Table } % Write "Table" before table number in list of tables
\setlength{\cfttabnumwidth}{6em} % Spacing before captions
\setlength{\cftbeforefigskip}{1ex} % Vertical spacing between entries

% Formatting the list of figures
\renewcommand{\listfigurename}{Supplemental figures} % Write "Supplemental figures" instead of "List of figures"
\renewcommand\cftfigpresnum{Figure } % Write "Figure" before fig number in list of figures
\setlength{\cftfignumwidth}{6em} % Spacing before captions
\setlength{\cftbeforetabskip}{1ex} % Vertical spacing between entries

\graphicspath{ {../figures/} } % tell Latex to look for figures in figures/

% Format figure and table numbers to follow S1, S2, ...
\renewcommand\thefigure{S\arabic{figure}} 
\renewcommand\thetable{S\arabic{table}} 

% Defining new environments as I will always want the figures centered
\newenvironment{cfigure}
	{\begin{figure} \centering}
	{\end{figure}}

\newenvironment{lsfigure}
	{\begin{landscape} \begin{figure} \centering}
	{\end{figure} \end{landscape}}

\setcounter{totalnumber}{1} % Allow a maximum of one float (table/figure) per page

\setlength{\parskip}{2ex}

% Information for the title page
\title{Supplemental data to ``Combined use of Oxford Nanopore and Illumina sequencing yields insights into soybean structural variation biology''}
\author{Marc-André Lemay \and Jonas A. Sibbesen \and Davoud Torkamaneh \and Jérémie Hamel \and Roger C. Levesque \and François Belzile}
\date{}

% END OF PREAMBLE

\begin{document}

\maketitle \thispagestyle{empty}

\tableofcontents

\vspace{4ex}

\listoftables

\thispagestyle{empty}
%\clearpage

\vspace{4ex}

\listoffigures

%\thispagestyle{empty}

\clearpage

\section{Description and testing of the SV refinement pipeline}

One of the main objectives of this project is to use Illumina data to genotype SVs that were originally discovered using Oxford Nanopore sequencing data. 
This would enable the use of Illumina data to carry out population-scale genotyping of variants that were initially discovered in a smaller set of samples sequenced with long-read technologies.

Oxford Nanopore data, however, has a high error rate.
This makes it challenging to accurately describe breakpoints and sequence content (for insertions) and may negatively impact our ability to genotype the variants discovered with this technology using short reads.

In order to address this issue, we assembled a pipeline that can be used to refine the SVs discovered by Sniffles \citep{ngmlr}.
By refinement, we mean that the position of the breakpoints (start and end position for deletions, and single location for insertions) and the sequence content (for insertions) can be updated to improve the quality of the variants. 
The current version of the pipeline ignores SVs other than deletions or insertions due to difficulties in assembling and aligning them properly.

The pipeline is based purely on Oxford Nanopore data (i.e. it does not need Illumina data e.g. for polishing or error correction) and therefore could potentially be used for SV refinement with any Oxford Nanopore or even PacBio dataset.

\subsection{Details of the steps in the pipeline}

The pipeline performs a local assembly of the Oxford Nanopore reads aligned to the region of interest and then uses AGE \citep{age} to align this assembly to the appropriate region of the reference genome.
This alignment is then used to update the breakpoints and inserted sequence of the variant if certain quality thresholds are met.

This alignment pipeline is largely inspired by the guidelines suggested by the authors of AGE in their newer LongAGE paper \citep{longage} and the authors of wtdbg2 \citep{wtdbg2} and consists of the following steps:

\begin{enumerate}
	\item For every insertion or deletion, Oxford Nanopore reads overlapping the region of interest are extracted from the NGMLR \citep{ngmlr} BAM file. 
		The region of interest is a window spanning ± 200 nucleotides from the SV breakpoints (start and end of the breakpoint for deletions, and single location for insertions).
\item  The reads extracted in this way are assembled using wtdbg2. 
	We use preset values for Oxford Nanopore sequencing (\texttt{-x ont}) except for the following parameters: \texttt{-S 1} (no subsampling of reads), \texttt{-e} 2 (minimum read coverage of 2 for an edge in the graph), \texttt{-L 1000} (minimum read size of 1 kb), \texttt{-g 500000} (expected genome size of 500 kb).
\item A consensus sequence for the assembly is produced using the \texttt{wtpoa-cns} command from wtdbg2 with default parameters.
\item This assembly is polished using Oxford Nanopore data. For this, the same reads that were used for assembly are aligned to the consensus sequence using minimap2 \citep{minimap2} with options \texttt{-ax map-ont -r2k}. These aligned reads are used to update the consensus sequence with \texttt{wtpoa-cns}.
\item Following assembly and polishing, the reference sequence is extracted ± 500 nucleotides from the SV breakpoints for the assembled contig to be aligned against it.
\item The largest of the assembled contigs (there is usually only one) is then aligned to the reference sequence using AGE with parameters \texttt{-allpos}, \texttt{-go=-1} (gap opening penalty of -1 to account for the high indel error rate of Oxford Nanopore data), \texttt{-indel} (the expected excised region is an indel) and \texttt{-both} (try alignment of both the sequence and its reverse complement, and report the best alignment).
\end{enumerate}

Once the alignment is completed, the results are parsed and used to extract the location of the regions excised by AGE from both the assembled contig and the reference genome.
In the case of a deletion, we expect a fragment of the reference genome to be excised, whereas in the case of a insertion, we expect a fragment of the assembled contig to be excised.
From this information, the location of the excised fragment is transformed to reference-based coordinates and the inserted sequence is extracted from the assembled contig (using the reverse complement if applicable).
The following metrics are then computed in order to compare the (refined) deletion/insertion suggested by the AGE alignment to the original (raw) one defined by Sniffles:

\begin{itemize}
	\item For deletions, the reciprocal overlap between the raw deletion and its refined counterpart. The reciprocal overlap is the minimum of the two following values: proportion of deletion 1 covered by deletion 2, and proportion of deletion 2 covered by deletion 1. This overlap is computed using functions from the GenomicRanges R package \citep{granges}.
	\item For insertions, the offset of the two insertions, i.e. the (absolute) physical distance in base pairs between the positions of the raw and refined insertion.
	\item For insertions, the Levenshtein (edit) distance between the two inserted sequences, as computed by the R function \texttt{adist}. From this distance, a {\em distance ratio} is computed, which corresponds to the Levensthein distance between the two insertions divided by the length of the longest insertion. This provides a relative measure of the difference in sequence content between the two insertions.
	\item In addition to the metrics related to the excised fragments themselves, the percentage of nucleotides that share identity and the percentage of nucleotides affected by gaps in the aligned flanking regions are also extracted from the output of AGE.
\end{itemize}


Note that SVs longer than 50 kb are excluded from the pipeline because of the high memory and computing time requirements of aligning to those variants using AGE.
Although a newer version of AGE called LongAGE \citep{longage} requiring less memory has been recently released, we found it unsuitable for our needs due to cryptic error messages and slower speed.

\subsection{Parameters determining whether SVs are updated or not}

Our objective is to update the position of the breakpoints and the sequence content such that the SVs updated in this way are closer to the actual sequence, and thus both easier to genotype from Illumina data and more representative of reality.
However, we must also avoid updating the variant in a way that is inaccurate (as would occur if, for example, AGE identified a deletion or insertion different from the target one).
Therefore, we established conditions to decide whether or not an SV should be updated.
Based on analyses carried out during the development of the pipeline, we have settled on the following thresholds to decide whether a variant should be updated:

\begin{itemize}
\item Reciprocal overlap $\geq$ 0.5 (for deletions)
\item Offset $\leq$ 50 (for insertions)
\item Edit distance ratio $\leq$ 0.5 (for insertions)
\item Percent identity of the alignment $\geq$ 85
\item Percentage of gaps in the assembly $\leq$ 15
\end{itemize}

\subsection{Testing the effect of refined SVs on genotyping performance}

The main objective of the SV refinement pipeline described above is to improve the genotyping of SVs with Illumina data.
In order to verify whether the pipeline indeed improved genotyping, we compared the genotyping precision and sensitivity of raw and refined SVs using three different genotypers: BayesTyper v. 1.5 \citep{bayestyper}, vg v. 1.23.0 \citep{vg} and Paragraph v. 2.4a \citep{paragraph}.

\subsubsection{Application of the pipeline}

For these testing purposes, we applied the SV refinement pipeline described above to the first four samples for which we obtained Oxford Nanopore data.
SVs discovered using Sniffles were filtered as described in the main text (notably, only SVs genotyped as homozygous were kept) and then used as input to the pipeline.
Insertions and deletions were refined if they met the required thresholds, which were as described above except for the minimum reciprocal overlap which was set to 0.7 and the maximum distance ratio which was set to 0.45.

Table \ref{table_s10} shows the number of variants that could be refined or not for each sample.
For all four samples, a majority of SVs subjected to the pipeline met the conditions for refinement.

There can be three main reasons why a variant is not refined:

\begin{itemize}
	\item The proposed refined variant did not meet one or several quality thresholds when compared to the raw variant.
	\item The size of the variant was $>$ 50 kb.
	\item No assembly or alignment was produced by the pipeline.
\end{itemize}

In either of these cases, we simply keep the original representation of the SV as called by Sniffles.

\subsubsection{Input SV datasets}

In order to assess the effect of refining SVs on genotpying performance, we generated an input dataset of raw variants and another input dataset of refined variants.

In order to generate input datasets for genotyping, the SV sets obtained from the four samples were merged using SVmerge \citep{svmerge} with parameters \texttt{-maxdist 15 -reldist 0.1 -relsizediff 0.1 -relshift 0.1 -seqspecific}.
For refined variants, the default random selection behavior of SVmerge when combining several variants together was modified to systematically favor variants that had been updated by the pipeline.

Following this, the VCF files were coordinate-sorted, normalized (using \texttt{bcftools norm}, \cite{bcftools}) and duplicate records were removed.

\subsubsection{Genotyping methods}

The raw and refined SV datasets were both used as input to three genotyping programs, BayesTyper, vg and Paragraph.

\paragraph{BayesTyper}

We followed the methods suggested by the authors of BayesTyper for genotyping (\url{https://github.com/bioinformatics-centre/BayesTyper}).

Briefly, k-mers were counted from the adapter-trimmed fastq files using KMC \citep{kmc} and a Bloom filter was applied using the \texttt{bayesTyperTools makeBloom} utility.

In parallel, \texttt{BayesTyperTools combine} had to be run to make the VCF input files suitable for input to BayesTyper.

Genotyping was then carried out using \texttt{bayesTyper cluster} and \texttt{bayesTyper genotype} sequentially with default parameters except for the use of the mitochondrial and chloroplastic genome sequences as decoy sequences.

Precision-recall curves for BayesTyper were generated by varying the genotype quality (GQ) threshold necessary to make a call.

\paragraph{vg}

Variation graphs were created from the VCF files using \texttt{vg construct} and then converted to PackedGraph format using \texttt{vg convert -p}. 
The graphs were then indexed prior to mapping using \texttt{vg index -L -x} to generate the xg index and \texttt{vg index -g} to generate the GCSA index. 
GCSA indexing was carried out on a version of the graph that had been pruned with \texttt{vg prune}.

Reads were mapped separately for each sample using \texttt{vg map}. 
These mappings were then used to call the SVs explicitly defined in the input VCF files using \texttt{vg pack -Q5} and \texttt{vg call -v}.

Precision-recall curves for vg were generated by varying the genotype quality (GQ) threshold necessary to make a call.

\paragraph{Paragraph}

VCF input files were prepared for input to Paragraph using similar methods to those described in the main text, i.e. variants located less than 500 nucleotides away from chromosome ends were removed and variants were padded with reference nucleotides where required. 
Samples were genotyped separately using the \texttt{multigrmpy} utility of Paragraph.

Precision-recall curves for Paragraph were generated by varying the minimum number of reads (DP) threshold necessary to make a call.

\subsubsection{Benchmarks: raw versus refined variants}

Benchmarks were carried out using the sveval R package \citep[][\url{https://github.com/jmonlong/sveval}]{vg} following the methods described in the main text.

\paragraph{Bayestyper}

Figure \ref{fig_s26} shows the benchmarking results obtained for BayesTyper.
Overall, the refined deletions resulted in a higher sensitivity but similar precision. 
The benefits of SV refinement were obvious for larger deletions, whereas results were sometimes worse for deletions in the range of 100-1000 nucleotides.
The improvemement brought by refinement is more obvious for insertions than it is for deletions, most likely because the refinement pipeline also improved the sequence content of the insertions.
The poor overall sensitivity of the results obtained with BayesTyper when genotyping Oxford Nanopore SVs is probably due to the small tolerance of this program to errors in SV breakpoint location or sequence content, as has been previously documented \citep{vg}.
Nevertheless, the improvement in sensitivity obtained by using refined SVs suggests that the SV refinement pipeline indeed improves the quality of the variants.

\paragraph{vg}

Figure \ref{fig_s27} shows the benchmarking results obtained for vg.
Refining the deletions does seem to have improved the results, although only very slightly.
For insertions, however, the improvement resulting from refinement is obvious, both in terms of sensitivity and precision.
Again, this suggests that the breakpoint refinement pipeline improved the accuracy of the sequence content of insertions, such that they could be more accurately genotyped by vg.

\paragraph{Paragraph}

Figure \ref{fig_s28} shows the benchmarking results obtained for Paragraph.
Using refined variants results in a slight increase in sensitivity for Paragraph, both for deletions and smaller insertions ($<$ 1,000 bp).
The apparent drop in precision for insertions $\geq$ 1,000 bp is likely due to the fact that the Oxford Nanopore data represented a poor ground truth dataset for these samples in this size range, as explained in the main text.
It is therefore expected that an increase in sensitivity from refining the variants will result in an apparent loss of precision in this size range.

\subsection{Conclusion on the SV refinement pipeline}

Overall, the results show that the breakpoint refinement pipeline results in higher genotyping sensitivity than using the raw variants as called by Sniffles.
While the results are not immediately obvious when using Paragraph, the improvements obtained using BayesTyper and vg suggest that the quality of the SVs is indeed higher following breakpoint refinement.
This is probably due to the high tolerance of Paragraph to errors in SV sequence (as documented in \cite{paragraph}) compared to the two other tools.
However, even if the improvement in genotyping performance using Paragraph is not obvious, downstream analyses should still benefit from more accurate SV breakpoints and sequence.

\subsection{Software availability}

The SV refinement pipeline described here is available from \url{https://github.com/malemay/breakpoint_refinement}.

\clearpage

\section{Detailed methods on the DNA transposable element analysis}

\subsection{Description of the analysis pipeline}

To study potential DNA transposable element (TE) activity in more detail, we extracted the SVs corresponding to all DNA TEs in the SoyTEdb database \citep{soytedb} that were matched by at least 3 SVs. 
This represented 51 TEs in the database and 237 individual SVs.

For each of these SVs, we extracted the Oxford Nanopore reads at the location of the SV and assembled them individually for each sample following methods similar to those used for the breakpoint refinement pipeline.
Briefly, reads overlapping positions ± 200 bp from the SV were used for assembly with wtdbg2 and polishing with wtpoa\_cns after re-aligning the reads to the assembly with minimap2.

The assemblies were then used for multiple alignment across samples in order to identify TE boundaries as well as target site duplication (TSD) and terminal inverted repeat (TIR) sequences.
We used the \texttt{PairwiseAlignment} function from the Bioconductor Biostrings package \citep{biostrings} to individually align all assemblies to a section of the reference genome ±~500 bp from the putative TE insertion site.
This pairwise alignment was only done in order to extract the part of the assemblies that was of interest.

We then used the aligned sequences obtained from the pairwise alignment as input for multiple alignment.
Multiple alignment was done with the MAFFT \citep{mafft} suite of tools using the command \texttt{ginsi --reorder} which performs multiple global alignment.

We processed multiple alignments so as to only keep high-quality alignments that showed obvious TE insertion polymorphism.
First, we manually checked all alignments and only kept those that showed an insertion/deletion polymorphism of the expected length.
Second, for a given multiple alignment, several of the sequences often showed very poor similarity to the reference genome, possibly because of issues with their assembly.
We removed such sequences by filtering out those that had less than 75\% percent identity with the reference in the first and last 500 bp of the alignment (the nucleotides flanking the insertion/deletion site).
Typically, only short TEs ($<$ 500 bp, essentially MITEs) remained following these filtering steps because the longer ones were too challenging to assemble and align properly.

A total of 72 SV events representing 25 individual entries in the TE database remained following filtering.
The sequences that remained following these filtering steps typically differed only in the presence or absence of the TE sequence.
Sequences that contained the TE sequence could be analyzed further to identify TSD and TIR sequences from the TE insertion sequence.
70 out of the 72 SVs had at least one sequence bearing the insertion after filtering.
In total, 301 sequences bearing the insertion could be queried for TSD and TIR sequences.
31 of these 301 were extracted from insertions that are present in the reference, whereas the 270 remaining insertion sequences were extracted from the local assemblies of the Oxford Nanopore reads.

We analyzed the insertion sequences using the GenericRepeatFinder software \citep{grf} with the following command:

\texttt{grf-main -i input\_fasta -o . -c 1 -p 30 -s 10 --seed\_mismatch 8 --min\_tr 12}

The parameters \texttt{--min\_space} and \texttt{--max\_space} were adjusted to be -60 and +20 relative to the expected SV size, respectively. 
The minimum (\texttt{--min\_tsd}) and maximum (\texttt{--max\_tsd}) TSD length were set to 2 for the DTT superfamily and 3 for the DTH superfamily, while these values were 5 and 10, respectively, for the DTM superfamily. 
We used relatively relaxed parameters for seed matching because the presence of TSD and TIR sequences was expected and we wanted to maximize the probability of finding them.

We parsed the output of GenericRepeatFinder to get the location and sequences of the TSD and TIR.
In many cases, several potential TSD/TIR sequences could be identified from a given input sequence. 
For DTT TEs, only entries with a TSD matching the expected TA sequence were kept. 
For DTH TEs, only entries with a TSD matching either TAA or TTA were kept. 
Despite those filters, several TSD/TIR candidates often remained for a given sample.
Therefore, the candidates were ranked following an automated procedure according to their proximity to the expected insertion location and similarity to expected insertion size.
For TEs in the DTM family, the highest priority was given to longer TSDs.
For every sample, we considered the TSD/TIR entry with the highest priority as the only match for that sample.

We annotated the multiple alignments with the putative TSD/TIR sequences and visually checked for consistency between these sequences and the location of the insertion.
Our analysis relied entirely on the automatically identified TSD/TIR sequences; samples for which the best matching TSD/TIR sequence was not consistent were discarded and were not analyzed further.
In total, we identified high-quality TSD/TIR sequences for 42 individual SVs.
For each of the TIR sequences identified, we computed the percentage of matching nucleotides between the two inverted repeats and averaged them over all insertion sequences for a given SV.

Some of the TIR sequences were identified directly from the reference sequence and may thus have been of higher quality than the error-prone assemblies generated from the Oxford Nanopore data.
Therefore, we only show the results obtained from the de novo-assembled sequences in figure 6c of the main text.
This results in 40 individual SVs collectively representing 17 sequences in the SoyTEdb database.

\subsection{Detailed analysis of a 480-bp Stowaway MITE insertion}

A closer look at the multiple alignments revealed that in almost all cases, the sequences without the insertion presented a single occurence of the TSD sequence.
This observation is consistent with a scenario where the TE never inserted into the sequence, instead of having excised from it.
In a single case however, that of a 480-bp insertion of a Tc1-Mariner (DTT) element (actually a Stowaway MITE) at position 2,257,090 of chromosome Gm04, visual analysis of the multiple alignment revealed that three different alleles are segregating at the insertion site:

\begin{itemize}
	\item The reference allele (no insertion at the target position)
	\item A 480-bp insertion that corresponds to the TE insertion
	\item A 6-bp insertion of nucleotides TACGAG 
\end{itemize}

We speculate that this last allele results from the excision of the TE insertion.
Indeed, the TA part of the insertion corresponds to the expected TSD sequence for a Tc1-Mariner element.
The remaining CGAG nucleotides would likely have been inserted during the repair of the DNA break following excision of the TE.
Interestingly, this insertion is by far the one for which the similarity between the two TIR sequences was highest among the ones studied, at 96.3\%.

To investigate this question further, we analysed the SNV patterns surrounding the insertion site.
If the TACGAG indeed resulted from the excision of the TE, then the haplotypes corresponding to the TE insertion and the 6-bp insertion should be more similar to each other than to the reference sequence, as the excision would have occurred within the TE insertion genetic background.

The SNVs with FILTER = PASS among those discovered by Platypus \citep{platypus} were extracted from the VCF file of the 102 samples sequenced by Illumina to perform this analysis.
The analysis of linkage disequilibrium (LD) using PLINK \citep{plink} in a 200-kb window surrounding the insertion location at Gm04:2,257,090 revealed a high-LD block spanning positions 2,220,398 to 2,259,326 that included the location of the insertion.
The SNVs in that 39-kb window were therefore extracted from the Platypus VCF file and used for downstream analyses.

By combining the information obtained from Platypus and SV genotyping with Paragraph, we were able to classify samples into three groups depending on their status at the TE insertion locus:

\begin{itemize}
	\item absent: Absence of the TE (reference allele)
	\item present: Presence of the TE insertion (as genotyped by Paragraph)
	\item excised: Presence of the 6-bp insertion putatively left by TE excision (as genotyped by Platypus)
\end{itemize}

The number of samples in each group was as follows:

\begin{itemize}
	\item absent: 71 
	\item present: 9
	\item excised: 14
\end{itemize}

All genotype calls for the TE insertion were homozygous and were therefore easy to classify as "present" when homozygous for the alternate allele.
For the 6-bp insertion, however, some genotype calls were heterozygous. 
In all such cases, visual analysis of the aligned Illumina reads confirmed that these were most likely mis-genotyped and were actually homozygous for the alternate allele.
Therefore, samples that were heterozygous at that site were classified as "excised" in addition to those that were homozygous for the alternate allele.
Samples that were called as homozygous for the reference allele at the 6-bp insertion position were classified as "absent".
Finally, eight (8) samples for which missing genotype calls did not allow unambiguous classification were discarded and not used for further analysis.

Using this classification, we computed the alternate allele frequency within each of the three allele classes of the 156 SNVs in the 39-kb LD block.
These results are visualized in figure 6d of the main text.
They show clear similarity between the "present" and "excised" haplotypes, which is indeed consistent with the 6-bp insertion being a remnant from the excision of the 480-bp TE.

\clearpage

\section{Subsampling analysis of four samples sequenced using Oxford Nanopore sequencing}

To investigate the effect of the relatively modest average sequencing depth (12X) of the Oxford Nanopore data used in this study, we performed a subsampling analysis of the data for four cultivars.

We adopted a subsampling approach in order to simulate lower coverage from the four cultivars that were sequenced with $>$ 16X sequencing depth (Alta, Maple Isle, AC2001, OAC Lakeview).
For each of the cultivars, we sampled 20\%, 40\%, 60\% and 80\% of the aligned reads using \texttt{samtools view --subsample} \citep{bcftools}.
Five (5) replicates of each cultivar/sampling fraction combination were generated using a different randomly chosen seed for each sample.
Each of the 80 subsampled \texttt{.bam} files were processed to generate a catalog of SV calls from each of them and benchmark their results using the same methods as described in the main text:

\begin{itemize}
	\item SV calling with Sniffles using parameters \texttt{--min\_support 3 --min\_seq\_size 1000 --min\_homo\_af 0.7}
	\item Filtering of variants by removing those meeting the following conditions:
		\begin{itemize}
			\item Variants other than insertions, deletions, duplications and inversions
			\item Variants called as heterozygous
			\item Variants overlapping any \texttt{N} in the reference sequence or containing any \texttt{N} in their sequence
			\item Variants located on unanchored scaffolds or organellar genomes
			\item Insertions located less than 20 bp away form any \texttt{N} in the reference sequence
			\item Variants smaller than 50 bp or larger than 500 kb
		\end{itemize}
	\item Benchmarking of each such filtered SV dataset with the sveval R package, using the SVs called on the complete (not sampled) dataset for the corresponding cultivar as a ground truth
\end{itemize}

Breakpoint refinement was not performed on the SV datasets generated from these subsamples for computational efficiency reasons.
For consistency, the complete (not sampled) truth SV sets against which subsampled datasets were compared were also not breakpoint-refined.
However, we believe that the impact of breakpoint refinement on these comparisons would have been minimal and should not impact the conclusions drawn from these analyses.

Results of the subsampling analyses are presented in figure \ref{fig_s7} (deletions), figure \ref{fig_s8} (insertions), figure \ref{fig_s9} (duplications) and figure \ref{fig_s10} (inversions).
In the context of this analysis, ``sensitivity'' represents the proportion of SVs in the complete dataset found in a given subsample.
Similarly, ``precision'' represents the proportion of SVs found in a given subsample that actually exist in the complete dataset.

The results obtained for deletions (figure \ref{fig_s7}) show that sensitivity exceeds 75\% at a sequencing depth of 10X across all size classes and samples.
The precision of deletion discovery is close to perfection across all size classes, and does not vary with sequencing depth.
These results indicate that, for deletions, a sequencing depth of 10X is sufficient to detect the majority of variants and that the occurrence of false positive calls does not increase in samples that are sequenced at low depth.

The results obtained for insertions (figure \ref{fig_s8}) show that sensitivity exceeds 75\% at a sequencing depth of 10X for insertions in the range 50 - 1,000 bp but decreases for larger insertions.
Therefore, it appears that the effect of sequencing depth becomes more important as insertions increase in size.
The precision of insertion discovery varies between 75\% and close to 100\% depending on insertion size, with the exception of insertions larger than 10 kb for a single sample (AC2001).
These results suggest that insertion discovery is less accurate than deletion discovery.
However, we observe the same pattern of independence between sequencing depth and precision as was observed for deletions, with the same exception as noted above.
This again indicates that the accuracy of insertion calls from Oxford Nanopore data does not decrease in samples that are sequenced at lower depth.

The results obtained for duplications (figure \ref{fig_s9}) show that duplication discovery is more sensitive to sequencing depth than other SV types, with sensitivity below 75\% at a sequencing depth of 10X.
This result is not surprising given that duplication calling relies on observing higher coverage on some genomic interval compared to average, and such differences can be assessed more confidently in deeply sequenced samples.
This result is also consistent with our observation that the apparent precision of duplication genotyping from Illumina data depends strongly on Oxford Nanopore sequencing depth (see figure \ref{fig_s6}).
The precision of duplication discovery seen in figure \ref{fig_s9} is also poorer than for other SV types and decreases slightly with decreasing sequencing depth, consistenly with difficulties in discovering and genotyping duplications as documented elsewhere in this paper.

Finally, the results obtained for inversions (figure \ref{fig_s10}) show sensitivity values similar to those observed for duplications, with sensitivity reaching between 50\% and 75\% at a sequencing depth of 10X.
Inversion discovery appears to be highly sensitive to sequencing depth, as the relationship between sensitivity and sequencing depth does not seem to plateau.
Precision rates are generally excellent ($>$ 90\%) and do not depend on sequencing depth, which indicates that inversion discovery will generally be accurate in samples sequenced at lower depth.
One exception to this result is smaller inversions (50 - 100 bp), as the small number of inversions observed in this size class (see table \ref{table_s1}) caused stochastic variation in precision rates depending on the cultivar and subsample.

The observation that the precision rates of deletion, insertion and inversion discovery do not vary with sequencing depth is not surprising.
Indeed, one would not expect a subsample of the reads to discover variants that were not already discovered from the full set of reads.
However, Sniffles determines the coordinates of SVs by combining the information from all reads supporting a given SV.
Therefore, for some variants located in regions where reads are difficult to align properly, subsets of reads may describe variants differently from the complete set of reads.
Similarly, variants that are supported by only a few misaligned reads might be called as homozygous from a sample of reads, but as heterozygous from the complete set and therefore filtered out.
Both these situations would lead to false positive calls in the subsampled dataset relative to the full dataset, as assessed by the sveval package.
Despite the possibility of those situations occurring, our results show that such false positive calls are rare for deletions, insertions, and inversions.
Therefore, it is safe to call these SVs from low-depth Oxford Nanopore sequencing data without worrying about the impact of false positive calls.

% Table S1 \caption[Number of SVs per Oxford Nanopore sample]{{{{

\begin{landscape}
\begin{table}
	\caption[Number of SVs per Oxford Nanopore sample]{
		Number of filtered variants per SV type and size class discovered by Sniffles for 17 samples subjected to Oxford Nanopore sequencing. 
		These SVs were used as a truth set for the benchmarks presented in this paper.}
	\label{table_s1}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		before first line = {\begin{tabular}{lccccccccccccccccccc}%
			\toprule & \multicolumn{4}{c}{[50bp -- 100bp[} & &  \multicolumn{4}{c}{[100bp -- 1kb[} & & \multicolumn{4}{c}{[1kb -- 10kb[} & & \multicolumn{4}{c}{$\geq$ 10kb}%
			\\[0.5ex] \cmidrule{2-5} \cmidrule{7-10} \cmidrule{12-15} \cmidrule{17-20}%
			& DEL\tnote{a} & INS\tnote{b} & DUP\tnote{c} & INV\tnote{d} & & DEL & INS & DUP & INV & & DEL & INS & DUP & INV & & DEL & INS & DUP & INV %
			\\[0.5ex] \cmidrule{2-5} \cmidrule{7-10} \cmidrule{12-15} \cmidrule{17-20}},%
			late after last line = { \\ \bottomrule \end{tabular}}]%
		{../tables/table_s1.csv}%
		{}%
		{\sample & \smalldel  & \smallins  & \smalldup  & \smallinv  & & %
		           \mediumdel & \mediumins & \mediumdup & \mediuminv & & %
		           \largedel  & \largeins  & \largedup  & \largeinv  & & %
			   \hugedel   & \hugeins   & \hugedup   & \hugeinv \\}

			   \begin{tablenotes}
			   \item[a] DEL: deletions
			   \item[b] INS: insertions
			   \item[c] DUP: duplications
			   \item[d] INV: inversions
			   \end{tablenotes}

	\end{threeparttable}

\end{table}
\end{landscape}

\clearpage%}}}

% Table S2 \caption{Number of deletions and insertions overlapping various genic features and proportion of these features within the reference annotation of Williams82}{{{

\begin{table}
	\centering
	\caption{Number of deletions and insertions overlapping various genic features and proportion of these features within the reference annotation of Williams82}
	\label{table_s2}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		tabular = lm{\widthof{\% in reference}}r@{\hskip 5pt}lr@{\hskip 5pt}lcm{\widthof{\% in reference}}r@{\hskip 2pt}lr@{\hskip 2pt}l,
		table head = \toprule & \multicolumn{5}{c}{Whole genome} & & \multicolumn{5}{c}{Non-repeated genome}\\ \cmidrule{2-6} \cmidrule{8-12}%
		Feature & \% in reference & DEL\tnote{a} & (\%) & INS\tnote{b} & (\%) & & \% in reference & DEL & (\%) & INS & (\%)\\\midrule,
		table foot = \bottomrule]%
		% DEPENDENCY: tables/table_s2.csv
		{../tables/table_s2.csv}%
		{}%
		{\feature & \centering \gprop & \DEL & (\DELprop) & \INS & (\INSprop) & &%
		\centering \nrgprop & \nrDEL & (\nrDELprop) & \nrINS & (\nrINSprop)\arraybackslash}

			   \begin{tablenotes}\footnotesize
			   \item[a] DEL: deletions
			   \item[b] INS: insertions
			   \item[c] cds: coding sequence
			   \item[d] upstream5kb: within 5 kb upstream of a gene
			   \end{tablenotes}

	\end{threeparttable}

\end{table}

\clearpage%}}}

% Table S3 \caption{Randomization test on the mean allele frequencies of SVs depending on the genic features overlapped}{{{

\begin{table}

	\centering
	\caption{Randomization test on the mean allele frequencies of SVs depending on the genic features overlapped}
	\label{table_s3}

	\begin{threeparttable}\small
		\csvreader[head to column names,
		before first line = {\begin{tabular}{l*5{c}}%
			\toprule & \multicolumn{2}{c}{Deletions} & & \multicolumn{2}{c}{Insertions} \\[0.5ex] \cmidrule{2-3} \cmidrule{5-6}%
			Genic features & Observed difference\tnote{a} & p-value\tnote{b} & & Observed difference & p-value \\\midrule},
		late after last line = {\\ \bottomrule \end{tabular}}]%
		% DEPENDENCY: tables/table_s3.csv
		{../tables/table_s3.csv}%
		{}%
		{\regionfmt & \deldiff & \delp & & \insdiff & \insp \\}

		\begin{tablenotes}\footnotesize
		\item[a] Observed difference between the mean allele frequencies of the SVs overlapping the corresponding genic features
		\item[b] p-value computed from a one-sided comparison to the distribution of mean differences obtained from the randomizations (10,000 iterations).
			The significance threshold was set to $\alpha = 0.05 \div 6 = 0.008$ to correct for multiple testing.
		\item[c] cds: coding sequence
		\item[d] upstream5kb: within 5 kb upstream of a gene
		\end{tablenotes}
	\end{threeparttable}

\end{table}

\clearpage%}}}

% Table S4 \caption{List of Gene Ontology Biological Process terms that are overrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}{{{

\begin{landscape}
\begin{table}

	\centering
	\caption{List of Gene Ontology Biological Process terms that are overrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}
	\label{table_s4}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		tabular = ll*4{c},
		table head = {\toprule GO ID & Description & Number of genes\tnote{a} & Expected\tnote{b} & Observed\tnote{c} & p-value\tnote{d} \\\midrule},
		table foot = {\bottomrule}]%
		% DEPENDENCY: tables/table_s4.csv
		{../tables/table_s4.csv}%
		{}%
		{\GOBPID & \Term & \Size & \ExpCount & \Count & \Pvalue}

		\begin{tablenotes}\footnotesize
		\item[a] Number of genes with corresponding GO annotation within the gene universe (whole genome except unanchored scaffolds)
		\item[b] Expected number of genes with given GO annotation within the subset of interest considering a hypergeometric distribution
		\item[c] Observed number of genes with given GO annotation within the subset of interest
		\item[d] Bonferroni-corrected p-values computed using the conditional hypergeometric test in the GOstats R package \citep{gostats}
		\end{tablenotes}
	\end{threeparttable}
\end{table}
\end{landscape}

\clearpage%}}}

% Table S5 \caption{List of Gene Ontology Biological Process terms that are underrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}{{{

\begin{landscape}
\begin{table}

	\centering
	\caption{List of Gene Ontology Biological Process terms that are underrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}
	\label{table_s5}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		tabular = ll*4{c},
		table head = {\toprule GO ID & Description & Number of genes\tnote{a} & Expected\tnote{b} & Observed\tnote{c} & p-value\tnote{d} \\\midrule},
		table foot = {\bottomrule}]%
		% DEPENDENCY: tables/table_s5.csv
		{../tables/table_s5.csv}%
		{}%
		{\GOBPID & \Term & \Size & \ExpCount & \Count & \Pvalue}

		\begin{tablenotes}\footnotesize
		\item[a] Number of genes with corresponding GO annotation within the gene universe (whole genome except unanchored scaffolds)
		\item[b] Expected number of genes with given GO annotation within the subset of interest considering a hypergeometric distribution
		\item[c] Observed number of genes with given GO annotation within the subset of interest
		\item[d] Bonferroni-corrected p-values computed using the conditional hypergeometric test in the GOstats R package \citep{gostats}
		\end{tablenotes}
	\end{threeparttable}
\end{table}
\end{landscape}

\clearpage%}}}

% Table S6 \caption{List of PFAM domains that are overrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}{{{

\begin{landscape}
\begin{table}

	\centering
	\caption{List of PFAM domains that are overrepresented ($\alpha = 0.05$) among coding sequences impacted by SVs with frequency $>$ 0.5}
	\label{table_s6}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		tabular = ll*4{c},
		table head = {\toprule PFAM ID & Description & Number of genes\tnote{a} & Expected\tnote{b} & Observed\tnote{c} & p-value\tnote{d} \\\midrule},
		table foot = {\bottomrule}]%
		% DEPENDENCY: tables/table_s6.csv
		{../tables/table_s6.csv}%
		{}%
		{\PFAMID & \Term & \Size & \ExpCount & \Count & \Pvalue}

		\begin{tablenotes}\footnotesize
		\item[a] Number of genes with corresponding PFAM annotation within the gene universe (whole genome except unanchored scaffolds)
		\item[b] Expected number of genes with given PFAM annotation within the subset of interest considering a hypergeometric distribution
		\item[c] Observed number of genes with given PFAM annotation within the subset of interest
		\item[d] Bonferroni-corrected p-values computed using the hypergeometric test in the GOstats R package \citep{gostats}
		\end{tablenotes}
	\end{threeparttable}
\end{table}
\end{landscape}

\clearpage%}}}

% Table S7 \caption{Metadata on the Illumina sequencing data of the 102 Canadian accessions}{{{
\begin{landscape} \small
	\begin{ThreePartTable}

		\begin{TableNotes}
		\item[a] The length of a single sequence read
		\item[b] The number of sequenced spots. Each spot corresponds to one forward and one reverse read.
		\item[c] The average sequencing depth computed from reads mapped to assembled chromosome pseudomolecules (Gm01 .. Gm20)
		\item[d] An "X" indicates that the sample belongs to the core set from which most Oxford Nanopore samples were selected
		\item[e] An "X" indicates that the sample was sequenced using Oxford Nanopore Technologies sequencing
		\end{TableNotes}

\begin{longtable}{llcccccl}

	\caption{Metadata on the Illumina sequencing data of the 102 Canadian accessions}\\

	\label{table_s7}\\

	\toprule
	ID & Name & Read length\tnote{a} & Number of spots\tnote{b} & Average depth\tnote{c} & Core\tnote{d} & ONT\tnote{e} & SRA Run accession \\
	\midrule \endfirsthead

	\caption[]{Metadata on the Illumina sequencing data of the 102 Canadian accessions \emph{(continued)}}\\

	\toprule
	ID & Name & Read length\tnote{a} & Number of spots\tnote{b} & Average depth\tnote{c} & Core\tnote{d} & ONT\tnote{e} & SRA Run accession \\
	\midrule
	\endhead

	\bottomrule
	Continued on next page \\
	\endfoot

	\bottomrule \insertTableNotes
	\endlastfoot
	\csvreader[head to column names,
		late after line=\\,
		late after last line=,
		before reading={\catcode`\#=12},
		after reading={\catcode`\#=6},
		/csv/separator=semicolon]%
	{../tables/table_s7.csv}{}{\ID & \name & \length & \spots & \depth & \core & \nanopore & \run}

\end{longtable}
\end{ThreePartTable}
\end{landscape}

\clearpage%}}}

% Table S8 \caption{DNA extraction and preparation methods for Oxford Nanopore sequencing}{{{
\begin{table}
	\centering
	\caption{DNA extraction and preparation methods for Oxford Nanopore sequencing}
	\label{table_s8}

	\begin{threeparttable}\small

		\csvreader[tabular = llllcll,
		head to column names,
		table head = \toprule Sample\tnote{a} & Growth\tnote{b} & Grinding\tnote{c} & %
			     Extraction\tnote{d} & RNase T (\textdegree{}C)\tnote{e} & Size selection\tnote{f} & Combined\tnote{g} \\\midrule,
		table foot = \bottomrule]%
% DEPENDENCY : table_s8.csv
		{../tables/table_s8.csv}%
		{}%
		{\sample & \growing & \grinding & \extraction & \rnase & \size & \combined}

		\begin{tablenotes}\footnotesize
		\item[a] The number in parentheses identifies samples sequenced on two different flowcells
		\item[b] Field: leaf harvested at Agriculture and Agri-Food Canada in Ottawa, and cryodessicated on site
		\item[c] Grinding method: LN + TL, Frozen in liquid nitrogen and ground using Qiagen TissueLyser;\\
			CD + TL, Cryodessicated at harvest and ground using Qiagen TissueLyser;\\
			LN + MP, Frozen in liquid nitrogen and ground using mortar and pestle
		\item[d] Extraction protocol used: CTAB, CTAB-based protocol described in main text;\\
			DNeasy, Qiagen DNeasy Plant Mini Kit;\\
			Gentra, Qiagen Gentra Puregene Tissue Kit
		\item[e] Incubation temperature of the RNase A treatment
		\item[f] Size-selection protocol: None, no size-selection;\\
			BP 6kb, BluePippin with 6-kb size-selection threshold;\\
			BP 15kb, BluePippin High-Pass Plus cassette with 15-kb size-selection threshold;\\
			SRE, Circulomics Short Read Eliminator Kit
		\item[g] Indicates whether the DNA from two different extractions was combined for sequencing
		\end{tablenotes}

	\end{threeparttable}
\end{table}

\clearpage%}}}

% Table S9 \caption{Metadata on Oxford Nanopore sequencing runs}{{{
\begin{table}
	\centering

	\caption{Metadata on Oxford Nanopore sequencing runs}
	\label{table_s9}

	\begin{threeparttable}\small
		\csvreader[tabular = llccccc,
		head to column names,
		table head = \toprule Sample\tnote{a} & Date\tnote{b} & Active pores\tnote{c} &%
		Mass (ng)\tnote{d} & Run time (h)\tnote{e} & Yield (Gb)\tnote{f} & Read N50 (kb)\tnote{g} \\\midrule,
		table foot = \bottomrule]%
% DEPENDENCY : table_s9.csv
		{../tables/table_s9.csv}%
		{}%
		{\sample & \date & \pores & \mass & \runtime & \yield & \nfifty}

		\begin{tablenotes}\footnotesize
		\item[a] The number in parentheses identifies samples sequenced on two different flowcells
		\item[b] Date when the sequencing run was started, format dd-mm-yyyy
		\item[c] Number of active pores on the flowcell prior to sequencing
		\item[d] Total estimated mass of the DNA library that was loaded on the flowcell
		\item[e] Duration of the sequencing run
		\item[f] Total number of bases in the fastq files following basecalling
		\item[g] Read N50 of all reads following basecalling

		\end{tablenotes}
	\end{threeparttable}
\end{table}

\clearpage%}}}

% Table S10 \caption{Classification of deletions and insertions for four samples used to test the SV refinement pipeline}{{{

\begin{table}
	\centering
	\caption{Classification of deletions and insertions for four samples used to test the SV refinement pipeline}
	\label{table_s10}

	\begin{threeparttable}\small

		\csvreader[head to column names,
		tabular = l*4{c},
		table head = {\toprule Sample & Refined\tnote{a} & Below thresholds\tnote{b} & Above max. SVLEN\tnote{c} & No alignment\tnote{d} \\\midrule},%
			table foot = {\bottomrule}]%
			% DEPENDENCY: tables/table_s10.csv
		{../tables/table_s10.csv}%
		{}%
		{\sample & \refined & \belowthr & \svlen & \noalignment}

		\begin{tablenotes}\footnotesize
		\item[a] SVs that passed all filters and had their breakpoints and/or sequence content updated by the pipeline
		\item[b] Did not pass one or several thresholds applied when comparing the proposed refined SV to the raw SV
		\item[c] SV length above the maximum permissible length (50 kb used here)
		\item[d] No assembly or no alignment produced by the pipeline
		\end{tablenotes}
	\end{threeparttable}

\end{table}

\clearpage%}}}

% Figure S1 \caption[Graphical summary of the methods used in this paper]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s1_diagram.png
	\includegraphics[width = 8.5in]{figure_s1_diagram}

	\flushleft{\emph{Caption on next page}}

	\label{fig_s1}
\end{lsfigure}

	\captionof{figure}[Graphical summary of the methods used in this paper]{
		Graphical summary of the methods used in this paper.
		The various steps of the analysis are color-coded depending on the data output by that step (see legend at the bottom of the figure).
		The words in italics in parentheses indicate the name of program used to compute that part of the analysis, where applicable.
		Arrows between steps indicate that one step is a prerequisite to the step it is pointing to.
		An exception to this is the dotted arrow indicating that both predicted SV genotypes and the ground truth Oxford Nanopore SV dataset are necessary to benchmarking.
		Benchmarks were performed for three different candidate SV datasets genotyped using Illumina data: SVs discoverered from Illumina data only, SVs discovered from Oxford Nanopore data only, and the combination of SVs discovered from the Illumina and Oxford Nanopore data.
	Population-scale analyses were performed on the SV dataset based on combined Illumina and Oxford Nanopore SVs.
	ONT: Oxford Nanopore Technologies, SV: structural variant, SNV: single-nucleotide variant, GO: Gene Ontology, BP: Biological Process, TIR: terminal inverted repeat, TSD: target site duplication, MITE: miniature inverted-repeat transposable element, PCA: principal component analysis.
}

\clearpage%}}}

% Figure S2 \caption[Heterozygosity rates of the population of 102 Canadian soybean cultivars]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s2.png
	\includegraphics[width = 3.5in]{figure_s2}

	\caption[Heterozygosity rates of the population of 102 Canadian soybean cultivars]{
		Heterozygosity rates of the population of 102 Canadian soybean cultivars as assessed from a set of filtered SNVs called by Platypus.
		(A) Distribution of heterozygosity rates of the 102 cultivars displayed as a histogram.
		(B) Relationship between the heterozygosity rate of a cultivar and its Illumina sequencing depth.
		The sequencing depth values shown are the same as in table \ref{table_s7} and represent the effective sequencing depth computed from mapped reads.
		With the exception of one outlier, heterozygosity rates higher than 0.1 are attributable to shallow sequencing, which can lead to genotyping errors resulting in seemingly higher heterozygosity.
		Heterozygous calls in the context of inbred lines may occur due to various causes including genotype-calling errors in low-depth samples (as shown here), copy number variants generating spurious heterozygous genotype calls \citep{torkamaneh2018}, or genuine residual heterozygosity.
	}

	\label{fig_s2}

\end{cfigure}

\clearpage%}}}

% Figure S3 \caption[Sensitivity and precision of Illumina deletion and insertion genotyping in non-repeat regions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s3.png
	\includegraphics[width = 5in]{figure_s3}

	\caption[Sensitivity and precision of Illumina deletion and insertion genotyping in non-repeat regions]{
		Genotyping sensitivity and precision of (A) deletions and (B) insertions discovered from the Illumina data in non-repeat regions. 
		Only SVs with less than 20\% overlap to regions annotated as repeats were used for these benchmarks. 
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
		Note that no results were available for deletions larger than 10 kb because these all overlapped repeat regions.
	}

	\label{fig_s3}

\end{cfigure}

\clearpage%}}}

% Figure S4 \caption[Sensitivity and precision of Illumina duplication and inversion genotyping]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s4.png
	\includegraphics[width = 3.5in]{figure_s4}
	
	\caption[Sensitivity and precision of Illumina duplication and inversion genotyping]{
		Genotyping sensitivity and precision of (A) duplications and (B) inversions discovered from the Illumina data.
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s4}

\end{cfigure}

\clearpage%}}}

% Figure S5 \caption[Sensitivity and precision of Illumina duplication and inversion genotyping in non-repeat regions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s5.png
	\includegraphics[width = 3.5in]{figure_s5}

	\caption[Sensitivity and precision of Illumina duplication and inversion genotyping in non-repeat regions]{
		Genotyping sensitivity and precision of (A) duplications and (B) inversions discovered from the Illumina data in non-repeat regions.
		Only SVs with less than 20\% overlap to regions annotated as repeats were used for these benchmarks. 
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s5}

\end{cfigure}

\clearpage%}}}

% Figure S6 \caption[Relationship between Oxford Nanopore sequencing depth and genotyping precision of Illumina SVs]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s6.png
	\includegraphics[width = 6in]{figure_s6}

	\caption[Relationship between Oxford Nanopore sequencing depth and genotyping precision of Illumina SVs]{
		Genotyping precision of (A) deletions, (B) insertions, (C) duplications and (D) inversions~discovered from Illumina data (at a minimum threshold of 2 supporting reads) for 17 samples as a function of the average Oxford Nanopore sequencing depth of that sample.
	}

	\label{fig_s6}

\end{cfigure}

\clearpage%}}}

% Figure s7 \caption[Effects of sequencing depth on the sensitivity and precision of deletion discovery from Oxford Nanopore sequencing data]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s7.png
	\includegraphics[width = 6.2in]{figure_s7}

	\caption[Effects of sequencing depth on the sensitivity and precision of deletion discovery from Oxford Nanopore sequencing data]{
		Effects of sequencing depth on the sensitivity and precision of deletion discovery from Oxford Nanopore sequencing data.
		Oxford Nanopore reads of four cultivars were subsampled using subsampling fractions of 20\%, 40\%, 60\% and 80\%, with five replicates of each cultivar/subsampling fraction combination.
		SVs were called on each of these subsamples and (A) sensitivity and (B) precision computed by comparing the resulting SVs to the SVs called on the full (not sampled) dataset for the corresponding cultivar.
		Points indicate the median value for a given cultivar/subsampling fraction combination, while vertical error bars indicate the span from the minimum to the maximum value of five replicates.
		Asterisks indicate the value (1 by definition) obtained for the full dataset.
		Error bars are not visible in cases where the minimum and maximum values are too close to each other.
}

\label{fig_s7}

\end{cfigure}

\clearpage%}}}

% Figure s8 \caption[Effects of sequencing depth on the sensitivity and precision of insertion discovery from Oxford Nanopore sequencing data]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s8.png
	\includegraphics[width = 6.2in]{figure_s8}

	\caption[Effects of sequencing depth on the sensitivity and precision of insertion discovery from Oxford Nanopore sequencing data]{
		Effects of sequencing depth on the sensitivity and precision of insertion discovery from Oxford Nanopore sequencing data.
		Oxford Nanopore reads of four cultivars were subsampled using subsampling fractions of 20\%, 40\%, 60\% and 80\%, with five replicates of each cultivar/subsampling fraction combination.
		SVs were called on each of these subsamples and (A) sensitivity and (B) precision computed by comparing the resulting SVs to the SVs called on the full (not sampled) dataset for the corresponding cultivar.
		Points indicate the median value for a given cultivar/subsampling fraction combination, while vertical error bars indicate the span from the minimum to the maximum value of five replicates.
		Asterisks indicate the value (1 by definition) obtained for the full dataset.
		Error bars are not visible in cases where the minimum and maximum values are too close to each other.
}

\label{fig_s8}

\end{cfigure}

\clearpage%}}}

% Figure s9 \caption[Effects of sequencing depth on the sensitivity and precision of duplication discovery from Oxford Nanopore sequencing data]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s9.png
	\includegraphics[width = 6.2in]{figure_s9}

	\caption[Effects of sequencing depth on the sensitivity and precision of duplication discovery from Oxford Nanopore sequencing data]{
		Effects of sequencing depth on the sensitivity and precision of duplication discovery from Oxford Nanopore sequencing data.
		Oxford Nanopore reads of four cultivars were subsampled using subsampling fractions of 20\%, 40\%, 60\% and 80\%, with five replicates of each cultivar/subsampling fraction combination.
		SVs were called on each of these subsamples and (A) sensitivity and (B) precision computed by comparing the resulting SVs to the SVs called on the full (not sampled) dataset for the corresponding cultivar.
		Points indicate the median value for a given cultivar/subsampling fraction combination, while vertical error bars indicate the span from the minimum to the maximum value of five replicates.
		Asterisks indicate the value (1 by definition) obtained for the full dataset.
		Error bars are not visible in cases where the minimum and maximum values are too close to each other.
}

\label{fig_s9}

\end{cfigure}

\clearpage%}}}

% Figure s10 \caption[Effects of sequencing depth on the sensitivity and precision of inversion discovery from Oxford Nanopore sequencing data]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s10.png
	\includegraphics[width = 6.2in]{figure_s10}

	\caption[Effects of sequencing depth on the sensitivity and precision of inversion discovery from Oxford Nanopore sequencing data]{
		Effects of sequencing depth on the sensitivity and precision of inversion discovery from Oxford Nanopore sequencing data.
		Oxford Nanopore reads of four cultivars were subsampled using subsampling fractions of 20\%, 40\%, 60\% and 80\%, with five replicates of each cultivar/subsampling fraction combination.
		SVs were called on each of these subsamples and (A) sensitivity and (B) precision computed by comparing the resulting SVs to the SVs called on the full (not sampled) dataset for the corresponding cultivar.
		Points indicate the median value for a given cultivar/subsampling fraction combination, while vertical error bars indicate the span from the minimum to the maximum value of five replicates.
		Asterisks indicate the value (1 by definition) obtained for the full dataset.
		Error bars are not visible in cases where the minimum and maximum values are too close to each other.
}

\label{fig_s10}

\end{cfigure}

\clearpage%}}}

% Figure S11 \caption[Effect of homozygous ALT count on genotyping performance of Illumina insertions and deletions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s11.png
	\includegraphics[width = 5in]{figure_s11}

	\caption[Effect of homozygous ALT count on genotyping performance of Illumina deletions and insertions]{
		Effect of filtering variants for minimum number of ALT alleles in homozygous genotype calls (homozygous ALT count) on the genotyping performance of (A) deletions and (B) insertions discovered from Illumina data.
		Each line and color represents a different sample.
		The curve for a given sample was obtained by varying the filtering threshold for a variant to be considered for benchmarking.
		The asterisks mark a homozygous ALT count threshold of 4; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s11}

\end{cfigure}

\clearpage%}}}

% Figure S12 \caption[Effect of homozygous ALT count on genotyping performance of Illumina duplications and inversions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s12.png
	\includegraphics[width = 3.5in]{figure_s12}

	\caption[Effect of homozygous ALT count on genotyping performance of Illumina duplications and inversions]{
		Effect of filtering variants for minimum number of ALT alleles in homozygous genotype calls (homozygous ALT count) on the genotyping performance of (A) duplications and (B) inversions discovered from Illumina data.
		Each line and color represents a different sample.
		The curve for a given sample was obtained by varying the filtering threshold for a variant to be considered for benchmarking.
		The asterisks mark a homozygous ALT count threshold of 6 (for duplications) or 10 (for inversions); points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s12}

\end{cfigure}

\clearpage%}}}

% Figure S13 \caption[Effect of the number of supporting calling tools on genotyping performance of Illumina deletions and insertions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s13.png
	\includegraphics[width = 5in]{figure_s13}

	\caption[Effect of the number of supporting calling tools on genotyping performance of Illumina deletions and insertions]{
		Effect of filtering variants for minimum number of supporting calling tools on the genotyping performance of (A) deletions and (B) insertions discovered from Illumina data.
		Each line and color represents a different sample.
		The curve for a given sample was obtained by varying the minimum threshold of number of supporting calling tools for a variant to be considered for benchmarking.
		The asterisks mark a minimum threshold of one supporting calling tool for a variant to be considered; points to the left of these for a given line represent increasingly stringent filtering threshold values.
		}

	\label{fig_s13}

\end{cfigure}

\clearpage%}}}

% Figure S14 \caption[Size distribution of the SVs reported by various calling tools using Illumina data]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s14.png
	\includegraphics[width = 6.5in]{figure_s14}

	\caption[Size distribution of the SVs reported by various calling tools using Illumina data]{
		Size distribution of the SVs reported by various calling tools using Illumina data.
		(A) Distribution of SV sizes in the range from -250 bp to 250 bp.
		Negative sizes represent deletions and positive sizes represent insertions.
		Duplications and inversions were omitted from this plot.
		(B) Distribution of the absolute size of deletions on a logarithmic scale.
	}

	\label{fig_s14}

\end{cfigure}

\clearpage%}}}

% Figure S15 \caption[Number of true positive and false positive deletion calls reported by various calling tool combinations]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s15.png
	\includegraphics[width = 9in]{figure_s15}

	\caption[Number of true positive and false positive deletion calls reported by various calling tool combinations]{
		Number of true positive and false positive deletion genotype calls broken down by size and by the combination of tools that reported the variant for sample OAC Oxford.
		OAC Oxford was chosen as a representative sample because of its median F1 score among all 17 benchmarked samples.
		The genotype calls considered here were filtered for a minimum number of supporting reads of 2 and a minimum homozygous ALT count of 4 across all 102 samples in the population.
		Note the logarithmic scale on the x-axis.
	}

	\label{fig_s15}

\end{lsfigure}

\clearpage%}}}

% Figure S16 \caption[Number of true positive and false positive insertion calls reported by various calling tool combinations]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s16.png
	\includegraphics[width = 6.5in]{figure_s16}

	\caption[Number of true positive and false positive insertion calls reported by various calling tool combinations]{
		Number of true positive and false positive insertion genotype calls broken down by size and by the combination of tools that reported the variant for sample OAC Lakeview.
		OAC Lakeview was chosen as a representative sample because of its median F1 score among all 17 benchmarked samples.
		The genotype calls considered here were filtered for a minimum number of supporting reads of 2 and a minimum homozygous ALT count of 4 across all 102 samples in the population.
		Note the logarithmic scale on the x-axis.
	}

	\label{fig_s16}

\end{cfigure}

\clearpage%}}}

% Figure S17 \caption[Relationship between Oxford Nanopore read N50 and genotyping precision of Oxford Nanopore insertions]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s17.png
	\includegraphics[width = 8in]{figure_s17}

	\caption[Relationship between Oxford Nanopore read N50 and genotyping precision of Oxford Nanopore insertions]{
		Genotyping precision of insertions discovered from Oxford Nanopore data (at a minimum threshold of 2 supporting reads) of 17 samples as a function of the read N50 of the Oxford Nanopore run(s) of that sample.
		(A) Precision of genotyping for insertions in the range 1,000 -- 10,000 bp.
		(B) Precision of genotyping for insertions $\geq$ 10,000 bp.

	}

	\label{fig_s17}
\end{lsfigure}

\clearpage%}}}

% Figure S18 \caption[Sensitivity and precision of Oxford Nanopore deletion and insertion genotyping in non-repeat regions]{{{{
\begin{lsfigure}
		% DEPENDENCY : figure_s18.png
		\includegraphics[width = 9.5in]{figure_s18}

		\caption[Sensitivity and precision of Oxford Nanopore deletion and insertion genotyping in non-repeat regions]{
			Genotyping sensitivity and precision of (A) deletions and (B) insertions discovered from the Oxford Nanopore data in non-repeat regions. 
			Only SVs with less than 20\% overlap to regions annotated as repeats were used for these benchmarks. 
			Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
			The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
			Note that no results were available for deletions larger than 10 kb because these all overlapped repeat regions.
	}

		\label{fig_s18}
\end{lsfigure}

\clearpage%}}}

% Figure S19 \caption[Sensitivity and precision of Oxford Nanopore duplication and inversion genotyping]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s19.png
	\includegraphics[width = 3.5in]{figure_s19}

	\caption[Sensitivity and precision of Oxford Nanopore duplication and inversion genotyping]{
		Genotyping sensitivity and precision of (A) duplications and (B) inversions discovered from the Oxford Nanopore data. 
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s19}

\end{cfigure}

\clearpage%}}}

% Figure S20 \caption[Sensitivity and precision of Oxford Nanopore duplication and inversion genotyping in non-repeat regions]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s20.png
	\includegraphics[width = 3.5in]{figure_s20}

	\caption[Sensitivity and precision of Oxford Nanopore duplication and inversion genotyping in non-repeat regions]{
		Genotyping sensitivity and precision of (A) duplications and (B) inversions discovered from the Oxford Nanopore data in non-repeat regions.
		Only SVs with less than 20\% overlap to regions annotated as repeats were used for these benchmarks. 
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s20}

\end{cfigure}

\clearpage%}}}

% Figure S21  \caption[Sensitivity and precision of joint Illumina/Oxford Nanopore deletion and insertion genotyping]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s21.png
	\includegraphics[width = 9.5in]{figure_s21}

	\caption[Sensitivity and precision of joint Illumina/Oxford Nanopore deletion and insertion genotyping]{
		Genotyping sensitivity and precision of (A) deletions and (B) insertions in the merged dataset of SVs discovered from the Illumina and Oxford Nanopore data.
			Each line and color represents one of 17 samples. 
			The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
			The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
	}

	\label{fig_s21}
\end{lsfigure}

\clearpage%}}}

% Figure S22 \caption[Sensitivity and precision of joint Illumina/Oxford Nanopore deletion and insertion genotyping in non-repeat regions]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s22.png
	\includegraphics[width = 9.5in]{figure_s22}

	\caption[Sensitivity and precision of joint Illumina/Oxford Nanopore deletion and insertion genotyping in non-repeat regions]{
		Genotyping sensitivity and precision of (A) deletions and (B) insertions in non-repeat regions in the merged dataset of SVs discovered from the Illumina and Oxford Nanopore data. 
		Only SVs with less than 20\% overlap to regions annotated as repeats were used for these benchmarks. 
		Each line and color represents one of 17 samples. 
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
		The asterisks indicate a minimum number of supporting reads of 2; points to the left of these for a given line represent increasingly stringent filtering threshold values.
		Note that no results were available for deletions larger than 10 kb because these all overlapped repeat regions.
	}

	\label{fig_s22}
\end{lsfigure}

\clearpage%}}}

% Figure S23 \caption[Principal component analyses (PCA) of the whole Canadian soybean panel]{{{{
\begin{cfigure}
	% DEPENDENCY : figure_s23.png
	\includegraphics[width = 6.5in]{figure_s23}

	\caption[Principal component analyses (PCA) of the whole Canadian soybean panel]{
		Principal component analyses (PCA) of the whole Canadian soybean panel.
		PCA were computed based on (A) SVs discovered from Illumina and Oxford Nanopore data, and subsequently genotyped with Illumina data using Paragraph and (B) SNVs called by Platypus from Illumina data.
		The first three principal components are shown for each PCA. 
		The color of each point indicates the population that the cultivar belongs to according to the fastStructure analysis with SNVs, following the same color scheme as in Figure 4 of the main text.
		Samples with a maximum q-value below 0.6 (as inferred from fastStructure) are considered admixed and are labeled with asterisks.
	}

	\label{fig_s23}

\end{cfigure}

\clearpage%}}}

% Figure s24 \caption[Neighbour-joining phylogenetic trees of the population of 102 Canadian soybean cultivars]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s24.png
	\includegraphics[width = 9.5in]{figure_s24}

	\caption[Neighbour-joining phylogenetic trees of the population of 102 Canadian soybean cultivars]{
		Neighbour-joining phylogenetic trees of the population of 102 Canadian soybean cultivars derived from (A) SVs discovered from Illumina and Oxford Nanopore data, and subsequently genotyped with Illumina data using Paragraph and (B) SNVs called by Platypus from Illumina data.
		Tip label colors indicate the population that the cultivar belongs to according to the fastStructure analysis with SNVs, following the same color scheme as in Figure 4 of the main text.
		Node labels indicate the number of trees (out of 100) of the bootstrap analysis that support a given node/grouping.
	}

	\label{fig_s24}
\end{lsfigure}

\clearpage%}}}

% Figure S25 \caption[IGV screenshot of Oxford Nanopore and Illumina read alignments of three samples at the location of a polymorphic Stowaway TE insertion]{{{
\begin{lsfigure}
	% DEPENDENCY : Gm04_2257090_annotated.png
	\includegraphics[width = 9.5in]{Gm04_2257090_annotated}

	\caption[IGV screenshot of Oxford Nanopore and Illumina read alignments of three samples at the location of a polymorphic Stowaway TE insertion]{
		IGV screenshot of Oxford Nanopore and Illumina read alignments of three samples at the location of a polymorphic Stowaway TE insertion at Gm04:2,257,090.
		(A) Oxford Nanopore and (B) Illumina alignments of sample CAD1077/AC2001, which matches the reference (absence of TE insertion).
		(C) Oxford Nanopore and (D) Illumina alignments of sample CAD1022/OAC Petrel, which bears the 480-bp Stowaway TE insertion.
		(E) Oxford Nanopore and (F) Illumina alignments of sample CAD1010/Maple Presto, which bears a 6-bp insertion putatively due to excision of the TE.
		The black rectangle in the middle of the figure encloses the region where the insertion polymorphisms of interest occur.
	}

	\label{fig_s25}

\end{lsfigure}

\clearpage%}}}

% Figure S26 \caption[Genotyping benchmark of raw and refined Oxford Nanopore SVs for four samples using BayesTyper]{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s26.png
	\includegraphics[width = 9.5in]{figure_s26}

	\caption[Genotyping sensitivity of raw and refined Oxford Nanopore SVs for four samples using BayesTyper]{
		Genotyping sensitivity and precision of raw and refined (A) deletions and (B) insertions discovered by Sniffles on the Oxford Nanopore data for four samples when using BayesTyper as a genotyper.
		Each line represents data for a single sample whose SV breakpoints have either been refined (refined SVs) or not (raw SVs).
		The points correspond to different filtering thresholds on the genotype quality (GQ) required to support a genotype call.
	}

	\label{fig_s26}

\end{lsfigure}

\clearpage%}}}

% Figure S27 \caption[Genotyping sensitivity of raw and refined Oxford Nanopore SVs for four samples using vg]{{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s27.png
	\includegraphics[width = 9.5in]{figure_s27}

	\caption[Genotyping sensitivity of raw and refined Oxford Nanopore SVs for four samples using vg]{
		Genotyping sensitivity and precision of raw and refined (A) deletions and (B) insertions discovered by Sniffles on the Oxford Nanopore data for four samples when using vg as a genotyper.
		Each line represents data for a single sample whose SV breakpoints have either been refined (refined SVs) or not (raw SVs).
		The points correspond to different filtering thresholds on the genotype quality (GQ) required to support a genotype call.
	}

	\label{fig_s27}

\end{lsfigure}

\clearpage%}}}

% Figure S28 \caption[Genotyping sensitivity of raw and refined Oxford Nanopore SVs for four samples using Paragraph]{{{
\begin{lsfigure}
	% DEPENDENCY : figure_s28.png
	\includegraphics[width = 9.5in]{figure_s28}

	\caption[Genotyping sensitivity of raw and refined Oxford Nanopore SVs for four samples using Paragraph]{
		Genotyping sensitivity and precision of raw and refined (A) deletions and (B) insertions discovered by Sniffles on the Oxford Nanopore data for four samples when using Paragraph as a genotyper.
		Each line represents data for a single sample whose SV breakpoints have either been refined (refined SVs) or not (raw SVs).
		The points correspond to different filtering thresholds on the minimum number of Illumina reads required to support a genotype call.
	}

	\label{fig_s28}

\end{lsfigure}

\clearpage%}}}

% DEPENDENCY : genome_research.bst
\bibliographystyle{vancouver}
% DEPENDENCY : references.bib
\addcontentsline{toc}{section}{References}
\bibliography{references}

\end{document}

